src/journal/block_map.rs

```diff
@@ -0,0 +1,215 @@
+
+        let Some(header) = JournalBlockHeader::read_bytes(&self.block)? else {
+            // Journal block magic is not present, so we've reached
+            // the end of the journal.
+            self.is_done = true;
+            return Ok(());
+        };
+
+        if header.sequence != self.sequence {
+            return Err(CorruptKind::JournalSequence.into());
+        }
+
+        if header.block_type == JournalBlockType::DESCRIPTOR {
+            self.process_descriptor_block()?;
+        } else if header.block_type == JournalBlockType::COMMIT {
+            self.process_commit_block()?;
+        } else {
+            return Err(IncompatibleKind::JournalBlockType(
+                header.block_type.0,
+            )
+            .into());
+        }
+        Ok(())
+    }
+
+    /// Process a descriptor block.
+    ///
+    /// Each descriptor block contains an array of tags, one for each
+    /// data block following the descriptor block. Each data block will
+    /// replace a block within the ext4 filesystem.
+    ///
+    /// Note that this will skip the `journal_block_iter` past the data
+    /// blocks that follow the descriptor block.
+    fn process_descriptor_block(&mut self) -> Result<(), Ext4Error> {
+        validate_descriptor_block_checksum(self.superblock, &self.block)?;
+
+        let tags = DescriptorBlockTagIter::new(&self.block[12..]);
+
+        for tag in tags {
+            let tag = tag?;
+
+            let block_index = self
+                .journal_block_iter
+                .next()
+                .ok_or(CorruptKind::JournalTruncated)??;
+
+            // Check the data block checksum.
+            let mut checksum = Checksum::new();
+            checksum.update(self.superblock.uuid.as_bytes());
+            checksum.update_u32_be(self.sequence);
+            self.fs
+                .read_from_block(block_index, 0, &mut self.data_block)?;
+            checksum.update(&self.data_block);
+            if checksum.finalize() != tag.checksum {
+                return Err(CorruptKind::JournalDescriptorTagChecksum.into());
+            }
+
+            self.uncommitted_block_map
+                .insert(tag.block_index, block_index);
+        }
+
+        Ok(())
+    }
+
+    /// Process a commit block.
+    ///
+    /// This indicates that a group of descriptor blocks have been
+    /// successfully processed. The entries in `uncommitted_block_map`
+    /// are moved to `block_map`, and the sequence number is
+    /// incremented.
+    fn process_commit_block(&mut self) -> Result<(), Ext4Error> {
+        validate_commit_block_checksum(self.superblock, &self.block)?;
+
+        // Commit the block map entries.
+        self.block_map.extend(self.uncommitted_block_map.iter());
+        self.uncommitted_block_map.clear();
+
+        self.sequence = self
+            .sequence
+            .checked_add(1)
+            .ok_or(CorruptKind::JournalSequenceOverflow)?;
```

Comment:
Have you tested how other things react to overflow? I wonder if an overflow is actually an error, or if it wraps. They aren't compared with `<`/`>`, so... or I guess they might be in the kernel.

I wonder how hard it would be to produce a test image that repro's a wrap. Might just need to reach in and twiddle the `s_sequence` in the journal superblock? Maybe `data=journal` described [here](https://www.kernel.org/doc/html/latest/filesystems/ext4/globals.html#journal-jbd2) would help more easily trigger the overflow?

Reply:
Good question, I have not checked the correct behavior. Filed https://github.com/nicholasbishop/ext4-view-rs/issues/434